sdk/samples/iot/provisioning/src/paho_iot_provisioning_example.c

```diff
@@ -321,135 +330,168 @@ static int register_device()
-    {
-      printf("Message from unknown topic, az_result return code %0x4\n", rc);
-      MQTTClient_freeMessage(&message);
-      MQTTClient_free(topic);
-      return rc;
-    }
-
-    printf("Received payload:\n");
-    print_az_span(message_span);
-    printf("Response status is %d.\n", response.status);
 
-    // Parse the operation status from a string to an enum
-    az_iot_provisioning_client_operation_status operation_status;
+    // Parse operation message
     if (az_failed(
-            rc = az_iot_provisioning_client_parse_operation_status(&response, &operation_status)))
+            rc = parse_operation_message(topic, topic_len, message, &response, &operation_status)))
     {
-      printf("Failed to parse operation_status, az_result return code %0x4\n", rc);
-      MQTTClient_freeMessage(&message);
-      MQTTClient_free(topic);
-      return rc;
+      printf("Failed to parse operation message, az_result return code %04x\n", rc);
+      return get_operation_status_free(rc, topic, message);
     }
+    printf("Parsed operation message.\n");
 
-    // Check whether or not the operation is complete
+    // If operation is not complete, send query and loop to receive operation message
     is_operation_complete = az_iot_provisioning_client_operation_complete(operation_status);
     if (!is_operation_complete)
     {
-      // In case the operation is not complete, issue a new query to the service
-      // Get the topic to send the query message
-      if (az_failed(
-              rc = az_iot_provisioning_client_query_status_get_publish_topic(
-                  &provisioning_client, &response, query_topic, sizeof(query_topic), NULL)))
-      {
-        printf("Unable to get query status publish topic, az_result return code %04x\n", rc);
-        MQTTClient_freeMessage(&message);
-        MQTTClient_free(topic);
-        return rc;
-      }
+      get_operation_status_free(0, topic, message);
 
-      // IMPORTANT: Wait the recommended retry-after number of seconds before query
-      printf("Querying after %u seconds...\n", response.retry_after_seconds);
-      sleep_for_seconds(response.retry_after_seconds);
-
-      // Publish the query message
-      if ((rc = MQTTClient_publish(mqtt_client, query_topic, 0, NULL, 0, 0, NULL))
-          != MQTTCLIENT_SUCCESS)
+      if ((rc = send_operation_query_message(response)) != MQTTCLIENT_SUCCESS)
       {
-        printf("Failed to publish query status request, MQTTClient return code %d\n", rc);
-        MQTTClient_freeMessage(&message);
-        MQTTClient_free(topic);
         return rc;
       }
-
-      // Loop back up to receive the result of the query
     }
-    else
+  } while (!is_operation_complete);
+
+  // Operation is complete
+  // Successful assignment - print out the assigned hostname and device id
+  if (operation_status == AZ_IOT_PROVISIONING_STATUS_ASSIGNED)
+  {
+    printf("SUCCESS - Device provisioned:\n");
+    printf("\tHub Hostname: ");
```

Comment:
possibly you can collapse some of printf

Reply:
Collapsed some.  Can improve on refactor in another PR.